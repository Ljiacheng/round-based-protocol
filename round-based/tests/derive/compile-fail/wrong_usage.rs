use round_based::ProtocolMessage;

#[derive(ProtocolMessage)]
enum Msg {
    // Unnamed variant with single field is the only correct enum variant
    // that doesn't contradicts with ProtocolMessage derivation
    VariantA(u16),
    // Error: You can't have named variants
    VariantB { n: u32 },
    // Error: Variant must have exactly 1 field
    VariantC(u32, String),
    // Error: Variant must have exactly 1 field!!
    VariantD(),
    // Error: Union variants are not permitted
    VariantE,
}

// Structure cannot implement ProtocolMessage
#[derive(ProtocolMessage)]
struct Msg2 {
    some_field: u64,
}

// Union cannot implement ProtocolMessage
#[derive(ProtocolMessage)]
union Msg3 {
    variant: u64,
}

fn main() {}
