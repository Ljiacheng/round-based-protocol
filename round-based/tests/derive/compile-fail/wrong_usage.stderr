error: named variants are not allowed in ProtocolMessage
 --> $DIR/wrong_usage.rs:9:5
  |
9 |     VariantB { n: u32 },
  |     ^^^^^^^^

error: this variant must contain exactly one field to be valid ProtocolMessage
  --> $DIR/wrong_usage.rs:11:5
   |
11 |     VariantC(u32, String),
   |     ^^^^^^^^

error: this variant must contain exactly one field to be valid ProtocolMessage
  --> $DIR/wrong_usage.rs:13:5
   |
13 |     VariantD(),
   |     ^^^^^^^^

error: unit variants are not allowed in ProtocolMessage
  --> $DIR/wrong_usage.rs:15:5
   |
15 |     VariantE,
   |     ^^^^^^^^

error: only enum may implement ProtocolMessage
  --> $DIR/wrong_usage.rs:20:1
   |
20 | struct Msg2 {
   | ^^^^^^

error: only enum may implement ProtocolMessage
  --> $DIR/wrong_usage.rs:26:1
   |
26 | union Msg3 {
   | ^^^^^
